### PR TITLE
C++ cleanup

### PR DESCRIPTION
**Pull request by bernardnormier**
This is a big C++ cleanup PR, with a particular focus on public header files.

The main goal was to fix the now-committed "local Slice" header files.

### CODE REVIEW

File: cpp/include/Ice/ServantLocator.h [hunk 1]
```diff
@@ -18,10 +17,10 @@ namespace Ice
      * @see ObjectAdapter#findServantLocator
      * \headerfile Ice/Ice.h
      */
-    class ICE_CLASS(ICE_API) ServantLocator
+    class ServantLocator
     {
     public:
-        ICE_MEMBER(ICE_API) virtual ~ServantLocator();
+        virtual ~ServantLocator() = default;
```
Comment (bernardnormier) [inline cpp/include/Ice/ServantLocator.h:23]: For the vast majority of our abstraction base classes, we never dynamic_cast to them so I believe this setup is fine.

One exception is NativePropertiesAdmin.

File: cpp/include/Ice/Communicator.h [hunk 2]
```diff
@@ -6,12 +6,13 @@
 #define ICE_COMMUNICATOR_H
 
 #include "Config.h"
+#include "CommunicatorF.h"
```
Comment (bernardnormier) [inline cpp/include/Ice/Communicator.h:9]: When we have a F.h file, we should include in the main .h file. This way, no need to redeclare the Ptr.

File: cpp/src/slice2cpp/CPlusPlusUtil.cpp [hunk 3]
```diff
@@ -510,7 +505,7 @@ Slice::typeToString(
     ClassDeclPtr cl = dynamic_pointer_cast<ClassDecl>(type);
     if (cl)
     {
-        return "::std::shared_ptr<" + getUnqualified(cl->scoped(), scope) + ">";
+        return getUnqualified(cl->scoped(), scope) + "Ptr";
```
Comment (bernardnormier) [inline cpp/src/slice2cpp/CPlusPlusUtil.cpp:508]: I updated the generated code to use/generate XxxPtr and not ::std::shared_ptr<Xxx>.

File: cpp/include/IceGrid/IceGrid.h [hunk 4]
```diff
@@ -5,22 +5,14 @@
...
-
-namespace IceGrid
-{
-    /**
-     * Obtains the plug-in facade for the IceGrid registry.
-     * @return The plug-in facade.
-     */
-    ICEGRID_API RegistryPluginFacadePtr getRegistryPluginFacade();
-}
+#include "Config.h"
```
Comment (bernardnormier) [inline cpp/include/IceGrid/IceGrid.h:8]: clang-format can sort the includes.

File: cpp/src/Ice/msbuild/ice/ice.vcxproj [hunk 5]
```diff
@@ -24,7 +24,7 @@
     <RootNamespace>Ice</RootNamespace>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
-  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='oDebug|Win32'" Label="Configuration">
```
Comment (pepone) [inline cpp/src/Ice/msbuild/ice/ice.vcxproj:27]: This seems like an unintended change.

File: cpp/src/IceGrid/PlatformInfo.h [hunk 6]
```diff
@@ -22,10 +22,7 @@ namespace IceGrid
     class PlatformInfo final
     {
     public:
-        PlatformInfo(
-            const std::string&,
-            const std::shared_ptr<Ice::Communicator>&,
-            const std::shared_ptr<TraceLevels>&);
+        PlatformInfo(const std::string&, const Ice::CommunicatorPtr&, const std::shared_ptr<TraceLevels>&);
```
Comment (pepone) [inline cpp/src/IceGrid/PlatformInfo.h:25]: No TraceLevelsPtr?

File: cpp/src/IceGrid/SessionI.h [hunk 7]
```diff
@@ -57,8 +57,7 @@ namespace IceGrid
     public:
         SessionI(const std::string&, const std::shared_ptr<Database>&, const IceUtil::TimerPtr&);
 
-        Ice::ObjectPrx
-        _register(const std::shared_ptr<SessionServantManager>&, const std::shared_ptr<Ice::Connection>&);
+        Ice::ObjectPrx _register(const std::shared_ptr<SessionServantManager>&, const Ice::ConnectionPtr&);
```
Comment (pepone) [inline cpp/src/IceGrid/SessionI.h:60]: We should add SessionServantManagerPtr for consistency

File: cpp/src/IceStorm/Grammar.cpp [hunk 8]
```diff
@@ -66,8 +66,8 @@
 // Copyright (c) ZeroC, Inc. All rights reserved.
 //
 
-#include <Ice/Ice.h>
-#include <IceStorm/Parser.h>
+#include "Ice/Ice.h"
+#include "Parser.h"
```
Comment (pepone) [inline cpp/src/IceStorm/Grammar.cpp:70]: This should be the same as in Grammar.y, we need to fix one of them.

File: cpp/src/IceStorm/Scanner.cpp [hunk 9]
```diff
@@ -407,9 +407,9 @@ char* yytext;
 // Copyright (c) ZeroC, Inc. All rights reserved.
 //
 
-#include <Ice/Ice.h>
-#include <IceStorm/Parser.h>
-#include <IceStorm/Grammar.h>
+#include "Ice/Ice.h"
+#include "IceStorm/Grammar.h"
+#include "Parser.h"
```
Comment (pepone) [inline cpp/src/IceStorm/Scanner.cpp:412]: This should match what we have in Scanner.l

**Review by pepone — APPROVED** [review]

File: cpp/include/Glacier2/SessionHelper.h [hunk 10]
```diff
@@ -31,10 +26,10 @@ namespace Glacier2
      * but better suited for graphical applications.
      * \headerfile Glacier2/Glacier2.h
      */
-    class GLACIER2_API SessionHelper
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Glacier2/SessionHelper.h:34]: We need XXX_API only when there is one or more out-of-line function to export / import (Windows) or make visible (non Windows).

SessionHelper no longer has any out of line function.

File: cpp/include/Ice/Initialize.h [hunk 11]
```diff
@@ -5,16 +5,16 @@
...
+#include "IceUtil/Timer.h"
+#include "InstanceF.h"
+#include "Instrumentation.h"
+#include "Logger.h"
 #include "Plugin.h"
-#include "BatchRequestInterceptor.h"
+#include "PropertiesF.h"
 #include "ValueFactory.h"
 
 #define ICE_CONFIG_FILE_STRING const std::string&
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Ice/Initialize.h:20]: Removed macro.

File: cpp/src/Ice/msbuild/ice/ice.vcxproj (hunk 5 above)
Comment (bernardnormier) [reply cpp/src/Ice/msbuild/ice/ice.vcxproj:27]: Yes, reverted, good catch!

File: cpp/src/IceGrid/PlatformInfo.h (hunk 6 above)
Comment (bernardnormier) [reply cpp/src/IceGrid/PlatformInfo.h:25]: More cleanup in follow-up PRs!

File: cpp/src/IceStorm/Grammar.cpp (hunk 8 above)
Comment (bernardnormier) [reply cpp/src/IceStorm/Grammar.cpp:70]: Fixed.

**Pull request merged**